ISPN-8024 Local Infinispan Hibernate cache tutorial for Wildfly

galderz opened the pull request:
Part of the work to add more Infinispan 2LC examples in [ISPN-8024](https://issues.jboss.org/browse/ISPN-8024).

To run, first start Wildfly 11 standalone. Then execute: 

    mvn clean package wildfly:deploy
    curl http://localhost:8080/wildfly-local/infinispan/hibernate-cache/entity
    curl http://localhost:8080/wildfly-local/infinispan/hibernate-cache/query
    curl http://localhost:8080/wildfly-local/infinispan/hibernate-cache/expiring

galderz commented:
Btw, I've used simple rest calls to emulate different hibernate/jpa operations. The reason I separated them is mostly because if all the operations are kept in same rest call, it wouldn't work since it will all be a single EntityManager session, and for example, querying would not work since an update happened in the same session (it compares timestamp when session is started and when the update happened). By having different endpoints, each call is a different session and hence it works. This seemed simpler than having to manage independent sessions within the example.

Btw, the order of the rest calls is currently expected the one above. I could name the rest calls as `/1`, `/2`, `/3` to make it more clearer.

If you any ideas on how to more easily show this, I'm all ears :)

slaskawi commented:
I'm not sure if I understand this correctly, but this PR depends on stuff in https://github.com/infinispan/infinispan-simple-tutorials/pull/33 ?

galderz commented:
It doesn't depend on anything. #33 is an example of a standalone Spring application. #32 is about a simple JAX-RS app running on Wildfly. Both are independent.

scottmarlow commented on hibernate-cache/wildfly-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/wildfly/local/InfinispanHibernateCacheWildflyLocal.java:
```diff
@@ -0,0 +1,247 @@
+
+   private void persistEntities() throws Exception {
+      clearStatistics();
+      userTransaction.begin();
```
After the userTransaction.begin(), you should call em.joinTransaction(), as I don't think that the container managed entity manager will otherwise be associated with the JTA transaction.

scottmarlow commented on hibernate-cache/wildfly-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/wildfly/local/InfinispanHibernateCacheWildflyLocal.java:
```diff
@@ -0,0 +1,247 @@
+
+   private void updateEntity(long id, StringBuilder out) throws Exception {
+      clearStatistics();
+      userTransaction.begin();
```
call em.joinTransaction() after userTransaction.begin()

scottmarlow commented on hibernate-cache/wildfly-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/wildfly/local/InfinispanHibernateCacheWildflyLocal.java:
```diff
@@ -0,0 +1,247 @@
+
+   private void deleteEntity(long id) throws Exception {
+      clearStatistics();
+      userTransaction.begin();
```
also call em.joinTransaction() after userTransaction.begin()

scottmarlow commented on hibernate-cache/wildfly-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/wildfly/local/InfinispanHibernateCacheWildflyLocal.java:
```diff
@@ -0,0 +1,247 @@
+
+   private void saveExpiringEntity() throws Exception {
+      clearStatistics();
+      userTransaction.begin();
```
also call em.joinTransaction() after userTransaction.begin()

galderz commented:
@scottmarlow I've changed the code so that no user transactions are needed. Essentially I've moved to numbered URLs that represent tutorial steps, and moved all the persistence code to an EJB which can manage the transactions for us.

galderz commented:
I'm integrating this. Feel free to add more comments if you find it's necessary and I'll address them.
